File: lib/services/youtube_service.dart


library;

import 'dart:convert';
import 'package:http/http.dart' as http;
import '../core/constants/app_constants.dart';

class YouTubeVideo {
  final String videoId;
  final String title;
  final String description;
  final String thumbnail;
  final String channelTitle;
  final String channelId;
  final String publishedAt;
  final String url;
  final int? viewCount;
  final int? likeCount;
  final String? duration;

  YouTubeVideo({
    required this.videoId,
    required this.title,
    required this.description,
    required this.thumbnail,
    required this.channelTitle,
    required this.channelId,
    required this.publishedAt,
    required this.url,
    this.viewCount,
    this.likeCount,
    this.duration,
  });

  factory YouTubeVideo.fromJson(Map<String, dynamic> json) {
    return YouTubeVideo(
      videoId: json['video_id'] ?? json['id']?['videoId'] ?? '',
      title: json['title'] ?? json['snippet']?['title'] ?? '',
      description: json['description'] ?? json['snippet']?['description'] ?? '',
      thumbnail: json['thumbnail'] ??
          json['snippet']?['thumbnails']?['medium']?['url'] ??
          '',
      channelTitle:
          json['channel_title'] ?? json['snippet']?['channelTitle'] ?? '',
      channelId: json['channel_id'] ?? json['snippet']?['channelId'] ?? '',
      publishedAt:
          json['published_at'] ?? json['snippet']?['publishedAt'] ?? '',
      url: json['url'] ??
          'https://www.youtube.com/watch?v=${json['video_id'] ?? json['id']?['videoId'] ?? ''}',
      viewCount: json['view_count'],
      likeCount: json['like_count'],
      duration: json['duration'],
    );
  }

  factory YouTubeVideo.fromYouTubeApi(Map<String, dynamic> item) {
    final snippet = item['snippet'] ?? {};
    final id = item['id'];
    final videoId = id is String ? id : (id?['videoId'] ?? '');

    return YouTubeVideo(
      videoId: videoId,
      title: snippet['title'] ?? '',
      description: snippet['description'] ?? '',
      thumbnail: snippet['thumbnails']?['medium']?['url'] ??
          snippet['thumbnails']?['default']?['url'] ??
          '',
      channelTitle: snippet['channelTitle'] ?? '',
      channelId: snippet['channelId'] ?? '',
      publishedAt: snippet['publishedAt'] ?? '',
      url: 'https://www.youtube.com/watch?v=$videoId',
    );
  }
}

class YouTubeChannel {
  final String name;
  final String id;
  final String language;

  YouTubeChannel({
    required this.name,
    required this.id,
    required this.language,
  });

  factory YouTubeChannel.fromJson(Map<String, dynamic> json) {
    return YouTubeChannel(
      name: json['name'],
      id: json['id'],
      language: json['language'],
    );
  }

  String get channelUrl => 'https://www.youtube.com/channel/$id';
}

class YouTubeService {
  static const String _baseUrl = 'https://www.googleapis.com/youtube/v3';
  static String get _apiKey => AppConstants.youtubeApiKey;

  static Future<Map<String, dynamic>> searchVideos({
    required String query,
    int maxResults = 10,
    String language = 'en',
    String duration = 'any',
    String order = 'relevance',
    bool useOnline = true,
  }) async {
    try {

      final searchQuery = '$query educational tutorial learning';

      final uri = Uri.parse('$_baseUrl/search').replace(queryParameters: {
        'part': 'snippet',
        'q': searchQuery,
        'type': 'video',
        'maxResults': maxResults.toString(),
        'order': order,
        'relevanceLanguage': language,
        'safeSearch': 'strict',
        'videoCategoryId': '27',
        'key': _apiKey,
      });

      final response = await http.get(uri).timeout(const Duration(seconds: 15));

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        final items = data['items'] as List? ?? [];

        final videos = items
            .where((item) => item['id']?['videoId'] != null)
            .map((item) => YouTubeVideo.fromYouTubeApi(item))
            .toList();

        return {
          'success': true,
          'provider': 'online',
          'videos': videos,
          'result_count': videos.length,
        };
      } else {

        return _getOfflineRecommendations(query);
      }
    } catch (e) {
      print('YouTube API error: $e');

      return _getOfflineRecommendations(query);
    }
  }

  static Map<String, dynamic> _getOfflineRecommendations(String query) {
    final subject = _detectSubject(query);
    final channels = _getChannelsForSubject(subject);

    return {
      'success': true,
      'provider': 'offline',
      'channels': channels,
      'detected_subject': subject,
      'note':
          'Showing recommended educational channels. Connect to internet for video search.',
    };
  }

  static String _detectSubject(String query) {
    final q = query.toLowerCase();
    if (q.contains('math') ||
        q.contains('algebra') ||
        q.contains('geometry') ||
        q.contains('calculus')) {
      return 'Mathematics';
    } else if (q.contains('physics') ||
        q.contains('force') ||
        q.contains('energy') ||
        q.contains('motion')) {
      return 'Physics';
    } else if (q.contains('chemistry') ||
        q.contains('chemical') ||
        q.contains('element') ||
        q.contains('reaction')) {
      return 'Chemistry';
    } else if (q.contains('biology') ||
        q.contains('cell') ||
        q.contains('organism') ||
        q.contains('life')) {
      return 'Biology';
    } else if (q.contains('history') ||
        q.contains('war') ||
        q.contains('civilization') ||
        q.contains('empire')) {
      return 'History';
    } else if (q.contains('english') ||
        q.contains('grammar') ||
        q.contains('literature')) {
      return 'English';
    } else if (q.contains('science')) {
      return 'Science';
    }
    return 'General';
  }

  static List<YouTubeChannel> _getChannelsForSubject(String subject) {
    final channelMap = {
      'Mathematics': [
        YouTubeChannel(
            name: '3Blue1Brown',
            id: 'UCYO_jab_esuFRV4b17AJtAw',
            language: 'English'),
        YouTubeChannel(
            name: 'Khan Academy',
            id: 'UC4a-Gbdw7vOaccHmFo40b9g',
            language: 'English'),
        YouTubeChannel(
            name: 'Mathologer',
            id: 'UC1_uAIS3r8Vu6JjXWvastJg',
            language: 'English'),
      ],
      'Physics': [
        YouTubeChannel(
            name: 'Veritasium',
            id: 'UCHnyfMqiRRG1u-2MsSQLbXA',
            language: 'English'),
        YouTubeChannel(
            name: 'Physics Wallah',
            id: 'UCrC79Pu5C3w_gxRf30_XMQQ',
            language: 'Hindi'),
        YouTubeChannel(
            name: 'MinutePhysics',
            id: 'UCUHW94eEFW7hkUMVaZz4eDg',
            language: 'English'),
      ],
      'Chemistry': [
        YouTubeChannel(
            name: 'NileRed',
            id: 'UCFhXFikryT4aFcLkLw2LBLA',
            language: 'English'),
        YouTubeChannel(
            name: 'Periodic Videos',
            id: 'UCtESv1e7ntJaLJYKIO1FoYw',
            language: 'English'),
        YouTubeChannel(
            name: 'Chemistry Wallah',
            id: 'UCrC79Pu5C3w_gxRf30_XMQQ',
            language: 'Hindi'),
      ],
      'Biology': [
        YouTubeChannel(
            name: 'CrashCourse',
            id: 'UCX6b17PVsYBQ0ip5gyeme-Q',
            language: 'English'),
        YouTubeChannel(
            name: 'Amoeba Sisters',
            id: 'UCkckDgmFpBxTZAYp0Q8VHDw',
            language: 'English'),
        YouTubeChannel(
            name: 'Biology Wallah',
            id: 'UCrC79Pu5C3w_gxRf30_XMQQ',
            language: 'Hindi'),
      ],
      'Science': [
        YouTubeChannel(
            name: 'Kurzgesagt',
            id: 'UCsXVk37bltHxD1rDPwtNM8Q',
            language: 'English'),
        YouTubeChannel(
            name: 'SmarterEveryDay',
            id: 'UC6107grRI4m0o2-emgoDnAA',
            language: 'English'),
        YouTubeChannel(
            name: 'Vsauce',
            id: 'UC6nSFpj9HTCZ5t-N3Rm3-HA',
            language: 'English'),
      ],
      'History': [
        YouTubeChannel(
            name: 'OverSimplified',
            id: 'UCNIuvl7V8zACPpTmmNIqP2A',
            language: 'English'),
        YouTubeChannel(
            name: 'Extra Credits',
            id: 'UCCODtTcd5M1JavPCOr_Uydg',
            language: 'English'),
        YouTubeChannel(
            name: 'History Matters',
            id: 'UC22BdTgxefuvUivrjevzjOw',
            language: 'English'),
      ],
      'English': [
        YouTubeChannel(
            name: 'English with Lucy',
            id: 'UCz4tgANd4yy8Oe0iXCdSWfA',
            language: 'English'),
        YouTubeChannel(
            name: 'Learn English with TV Series',
            id: 'UCKgpamMlm872zkGDcBJHYDg',
            language: 'English'),
      ],
      'General': [
        YouTubeChannel(
            name: 'Khan Academy',
            id: 'UC4a-Gbdw7vOaccHmFo40b9g',
            language: 'English'),
        YouTubeChannel(
            name: 'TED-Ed',
            id: 'UCsooa4yRKGN_zEE8iknghZA',
            language: 'English'),
        YouTubeChannel(
            name: 'CrashCourse',
            id: 'UCX6b17PVsYBQ0ip5gyeme-Q',
            language: 'English'),
      ],
    };

    return channelMap[subject] ?? channelMap['General']!;
  }

  static Future<YouTubeVideo?> getVideoDetails(String videoId) async {
    try {
      final uri = Uri.parse('$_baseUrl/videos').replace(queryParameters: {
        'part': 'snippet,statistics,contentDetails',
        'id': videoId,
        'key': _apiKey,
      });

      final response = await http.get(uri).timeout(const Duration(seconds: 10));

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        final items = data['items'] as List? ?? [];

        if (items.isNotEmpty) {
          final item = items.first;
          final snippet = item['snippet'] ?? {};
          final stats = item['statistics'] ?? {};
          final contentDetails = item['contentDetails'] ?? {};

          return YouTubeVideo(
            videoId: videoId,
            title: snippet['title'] ?? '',
            description: snippet['description'] ?? '',
            thumbnail: snippet['thumbnails']?['high']?['url'] ?? '',
            channelTitle: snippet['channelTitle'] ?? '',
            channelId: snippet['channelId'] ?? '',
            publishedAt: snippet['publishedAt'] ?? '',
            url: 'https://www.youtube.com/watch?v=$videoId',
            viewCount: int.tryParse(stats['viewCount']?.toString() ?? ''),
            likeCount: int.tryParse(stats['likeCount']?.toString() ?? ''),
            duration: contentDetails['duration'],
          );
        }
      }
      return null;
    } catch (e) {
      print('Error getting video details: $e');
      return null;
    }
  }

  static Future<List<YouTubeChannel>> getChannelRecommendations({
    String subject = 'General',
  }) async {
    return _getChannelsForSubject(subject);
  }

  static Future<bool> checkHealth() async {
    try {
      final uri = Uri.parse('$_baseUrl/videos').replace(queryParameters: {
        'part': 'snippet',
        'chart': 'mostPopular',
        'maxResults': '1',
        'key': _apiKey,
      });

      final response = await http.get(uri).timeout(const Duration(seconds: 5));
      return response.statusCode == 200;
    } catch (e) {
      print('YouTube health check failed: $e');
      return false;
    }
  }

  static List<String> get subjects => [
        'General',
        'Science',
        'Mathematics',
        'English',
        'History',
        'Physics',
        'Chemistry',
        'Biology',
      ];
}

File: backend/services/youtube_service.py
"""
YouTube Data API Service
Provides video search and recommendations for educational content
"""

import os
from typing import Dict, Any, Optional, List
from datetime import datetime, timedelta

class YouTubeService:
    """
    YouTube Data API v3 service for content discovery
    - Search videos by topic/subject
    - Get video details
    - Filter educational content
    - Curated channel recommendations (offline fallback)
    """

    def __init__(self, api_key: Optional[str] = None):
        """
        Initialize YouTube service

        Args:
            api_key: YouTube Data API v3 key (optional for offline mode)
        """
        self.api_key = api_key
        self.online_available = api_key is not None

        self.educational_channels = {
            'Science': [
                {'name': 'Khan Academy', 'id': 'UC4a-Gbdw7vOaccHmFo40b9g', 'language': 'en'},
                {'name': 'Crash Course', 'id': 'UCX6b17PVsYBQ0ip5gyeme-Q', 'language': 'en'},
                {'name': 'Veritasium', 'id': 'UCHnyfMqiRRG1u-2MsSQLbXA', 'language': 'en'},
                {'name': 'Physics Wallah', 'id': 'UCeVMnSShP_Iviwkknt83cww', 'language': 'hi'},
            ],
            'Mathematics': [
                {'name': 'Khan Academy', 'id': 'UC4a-Gbdw7vOaccHmFo40b9g', 'language': 'en'},
                {'name': '3Blue1Brown', 'id': 'UCYO_jab_esuFRV4b17AJtAw', 'language': 'en'},
                {'name': 'Vedantu', 'id': 'UCHEg7jlYU-BFC8n_XKgHqbQ', 'language': 'hi'},
            ],
            'English': [
                {'name': 'EngVid', 'id': 'UCpvuCyYG3-', 'language': 'en'},
                {'name': 'BBC Learning English', 'id': 'UCHaHD477h-FeBbVh9Sh7syA', 'language': 'en'},
            ],
            'History': [
                {'name': 'Crash Course', 'id': 'UCX6b17PVsYBQ0ip5gyeme-Q', 'language': 'en'},
                {'name': 'Oversimplified', 'id': 'UCNIuvl7V8zACPpTmmNIqP2A', 'language': 'en'},
            ],
            'General': [
                {'name': 'Khan Academy', 'id': 'UC4a-Gbdw7vOaccHmFo40b9g', 'language': 'en'},
                {'name': 'CrashCourse', 'id': 'UCX6b17PVsYBQ0ip5gyeme-Q', 'language': 'en'},
                {'name': 'Unacademy', 'id': 'UCABe2FgVNv2hgBeMu2mySVg', 'language': 'hi'},
                {'name': "Byju's", 'id': 'UCF2JsMDcfUPcvMkTmx_BzfQ', 'language': 'hi'},
            ]
        }

    def search_videos(
        self,
        query: str,
        max_results: int = 10,
        language: str = 'en',
        duration: str = 'any',
        order: str = 'relevance',
        use_online: bool = True
    ) -> Dict[str, Any]:
        """
        Search for educational videos

        Args:
            query: Search query
            max_results: Maximum results to return
            language: Preferred language (en, hi, etc.)
            duration: Video duration filter (short/medium/long/any)
            order: Sort order (relevance/date/rating/viewCount)
            use_online: Use YouTube API if available

        Returns:
            Dictionary with video results
        """
        if not query or not query.strip():
            return {
                'success': False,
                'error': 'Empty search query'
            }

        if use_online and self.online_available:
            return self._search_youtube_api(query, max_results, language, duration, order)

        return self._search_offline(query, max_results)

    def _search_youtube_api(
        self,
        query: str,
        max_results: int,
        language: str,
        duration: str,
        order: str
    ) -> Dict[str, Any]:
        """
        Search using YouTube Data API

        Args:
            query: Search query
            max_results: Maximum results
            language: Language preference
            duration: Duration filter
            order: Sort order

        Returns:
            Dictionary with video results
        """
        try:

            import requests

            url = "https://www.googleapis.com/youtube/v3/search"

            duration_map = {
                'short': 'short',
                'medium': 'medium',
                'long': 'long',
                'any': None
            }

            params = {
                'key': self.api_key,
                'part': 'snippet',
                'q': f"{query} tutorial educational",
                'type': 'video',
                'maxResults': max_results,
                'order': order,
                'relevanceLanguage': language,
                'videoDefinition': 'any',
                'videoEmbeddable': 'true',
                'safeSearch': 'strict'
            }

            if duration != 'any' and duration in duration_map:
                params['videoDuration'] = duration_map[duration]

            response = requests.get(url, params=params, timeout=10)

            if response.status_code == 200:
                result = response.json()

                videos = []
                for item in result.get('items', []):
                    snippet = item.get('snippet', {})
                    video_id = item.get('id', {}).get('videoId')

                    if video_id:
                        videos.append({
                            'video_id': video_id,
                            'title': snippet.get('title', ''),
                            'description': snippet.get('description', ''),
                            'thumbnail': snippet.get('thumbnails', {}).get('medium', {}).get('url', ''),
                            'channel_title': snippet.get('channelTitle', ''),
                            'channel_id': snippet.get('channelId', ''),
                            'published_at': snippet.get('publishedAt', ''),
                            'url': f"https://www.youtube.com/watch?v={video_id}"
                        })

                return {
                    'success': True,
                    'provider': 'youtube_api',
                    'query': query,
                    'result_count': len(videos),
                    'videos': videos
                }
            else:
                error_msg = response.json().get('error', {}).get('message', 'Unknown error')
                return {
                    'success': False,
                    'error': f'YouTube API error: {error_msg}',
                    'fallback_to_offline': True
                }

        except ImportError:
            return {
                'success': False,
                'error': 'requests library not installed',
                'fallback_to_offline': True
            }
        except Exception as e:
            return {
                'success': False,
                'error': f'YouTube search failed: {str(e)}',
                'fallback_to_offline': True
            }

    def _search_offline(self, query: str, max_results: int) -> Dict[str, Any]:
        """
        Provide offline recommendations based on query

        Args:
            query: Search query
            max_results: Maximum results

        Returns:
            Dictionary with channel recommendations
        """

        subject = self._detect_subject(query)

        channels = self.educational_channels.get(subject, self.educational_channels['General'])

        return {
            'success': True,
            'provider': 'offline',
            'query': query,
            'detected_subject': subject,
            'recommended_channels': channels[:max_results],
            'note': 'Offline mode: Showing curated educational channels. Use online mode for specific video search.'
        }

    def _detect_subject(self, query: str) -> str:
        """Detect subject from query"""
        query_lower = query.lower()

        subject_keywords = {
            'Science': ['science', 'physics', 'chemistry', 'biology', 'photosynthesis', 'experiment'],
            'Mathematics': ['math', 'mathematics', 'algebra', 'geometry', 'calculus', 'equation'],
            'English': ['english', 'grammar', 'vocabulary', 'writing', 'literature'],
            'History': ['history', 'historical', 'ancient', 'medieval', 'war', 'civilization']
        }

        for subject, keywords in subject_keywords.items():
            if any(keyword in query_lower for keyword in keywords):
                return subject

        return 'General'

    def get_video_details(self, video_id: str) -> Dict[str, Any]:
        """
        Get detailed information about a video

        Args:
            video_id: YouTube video ID

        Returns:
            Dictionary with video details
        """
        if not self.online_available:
            return {
                'success': False,
                'error': 'YouTube API not available (offline mode)',
                'video_id': video_id
            }

        try:
            import requests

            url = "https://www.googleapis.com/youtube/v3/videos"
            params = {
                'key': self.api_key,
                'part': 'snippet,contentDetails,statistics',
                'id': video_id
            }

            response = requests.get(url, params=params, timeout=10)

            if response.status_code == 200:
                result = response.json()

                if result.get('items'):
                    item = result['items'][0]
                    snippet = item.get('snippet', {})
                    stats = item.get('statistics', {})
                    details = item.get('contentDetails', {})

                    return {
                        'success': True,
                        'video_id': video_id,
                        'title': snippet.get('title', ''),
                        'description': snippet.get('description', ''),
                        'channel_title': snippet.get('channelTitle', ''),
                        'published_at': snippet.get('publishedAt', ''),
                        'duration': details.get('duration', ''),
                        'view_count': stats.get('viewCount', 0),
                        'like_count': stats.get('likeCount', 0),
                        'comment_count': stats.get('commentCount', 0),
                        'thumbnail': snippet.get('thumbnails', {}).get('high', {}).get('url', ''),
                        'url': f"https://www.youtube.com/watch?v={video_id}"
                    }
                else:
                    return {
                        'success': False,
                        'error': 'Video not found'
                    }
            else:
                return {
                    'success': False,
                    'error': 'Failed to fetch video details'
                }

        except Exception as e:
            return {
                'success': False,
                'error': str(e)
            }

    def get_channel_recommendations(self, subject: str = 'General') -> List[Dict[str, str]]:
        """
        Get curated channel recommendations

        Args:
            subject: Subject area

        Returns:
            List of recommended channels
        """
        return self.educational_channels.get(subject, self.educational_channels['General'])

    def health_check(self) -> Dict[str, str]:
        """Check YouTube service health"""
        return {
            'youtube_service': 'healthy',
            'youtube_api': 'available' if self.online_available else 'unavailable',
            'offline_channels': len([ch for channels in self.educational_channels.values() for ch in channels])
        }

_youtube_service = None

def get_youtube_service(api_key: Optional[str] = None) -> YouTubeService:
    """Get singleton YouTube service instance"""
    global _youtube_service
    if _youtube_service is None:
        _youtube_service = YouTubeService(api_key)
    return _youtube_service
